pkg/deploy/service_accounts.go

```diff
@@ -52,6 +55,66 @@ func (s *ServiceAccounts) Find(genericOpts GenericOpts, saName string) (Processe
 	return pgoForSA, nil
 }
 
+// GetClusterVersion returns the kubernetes API version for the cluster which has been supplied to kapp-controller via a kubeconfig
+func GetClusterVersion(cc kubernetes.Interface, saName string, specCluster *v1alpha1.AppCluster, objMeta *metav1.ObjectMeta, log logr.Logger) (*version.Info, error) {
+	// this logic is duplicated here and in ProcessOpts: if the serviceAccount name is present, we will be deploying to this cluster
+	if len(saName) > 0 {
+		return cc.Discovery().ServerVersion()
+	}
+
+	processedGenericOpts, err := processOpts(
+		saName,
+		specCluster,
+		GenericOpts{Name: objMeta.Name, Namespace: objMeta.Namespace},
+		NewServiceAccounts(cc, log),
+		NewKubeconfigSecrets(cc))
```

Comment:
service accounts object should be shared with the one in the factory to take advantage of token cache.